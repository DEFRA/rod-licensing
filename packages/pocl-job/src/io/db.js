import config from '../config.js'
import { docClient } from '../../../connectors-lib/src/aws.js'
import { GetCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb'

/**
 * Update the POCL file staging table to add or update the entry for the provided filename
 *
 * @param {string} filename the primary key of the record in the file staging table to add/update
 * @param {Object} entries to update the table with
 * @returns {Promise<void>}
 */
export const updateFileStagingTable = async ({ filename, ...entries }) => {
  const updateExpression = docClient.createUpdateExpression({
    expires: Math.floor(Date.now() / 1000) + config.db.stagingTtlDelta,
    ...entries
  })
  await docClient.send(
    new UpdateCommand({
      TableName: config.db.fileStagingTable,
      Key: { filename },
      ...updateExpression
    })
  )
}

/**
 * Retrieve all file records for the specified stages.  If stages are not provided then all records are returned
 *
 * @param {string} stages the stage names to filter the result-set on. Omit to retrieve all records regardless of the stage.
 * @returns {Promise<[DocumentClient.AttributeMap]>}
 */
export const getFileRecords = async (...stages) => {
  const stageValues = stages.reduce((acc, s, i) => ({ ...acc, [`:stage${i}`]: s }), {})
  return docClient.scanAllPromise({
    TableName: config.db.fileStagingTable,
    ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues)})` }),
    ExpressionAttributeValues: stageValues,
    ConsistentRead: true
  })
}

/**
 * Retrieve an individual file record by the specified filename
 *
 * @param filename the name of the POCL file to retrieve a record for
 * @returns {DocumentClient.AttributeMap}
 */
export const getFileRecord = async filename => {
  const result = await docClient.send(
    new GetCommand({
      TableName: config.db.fileStagingTable,
      Key: { filename },
      ConsistentRead: true
    })
  )
  return result.Item
}

/**
 * Update the POCL record staging table to add entries for each of the provided records
 *
 * @param {string} filename the filename of a POCL file to which the records relate
 * @param {Array<Object>} records to update the table with
 * @returns {Promise<void>}
 */
export const updateRecordStagingTable = async (filename, records) => {
  if (records.length) {
    const params = {
      RequestItems: {
        [config.db.recordStagingTable]: records.map(record => ({
          PutRequest: { Item: { filename, expires: Math.floor(Date.now() / 1000) + config.db.stagingTtlDelta, ...record } }
        }))
      }
    }
    await docClient.batchWriteAllPromise(params)
  }
}

/**
 * Retrieve processed records from the record staging table for the provided filename and stage filters
 *
 * @param {string} filename the filename of a POCL file for which the records should be retrieved
 * @param {string} stages the stage names to filter the result-set on. Omit to retrieve all records regardless of the stage.
 * @returns {Promise<[DocumentClient.AttributeMap]>}
 */
export const getProcessedRecords = async (filename, ...stages) => {
  const stageValues = stages.reduce((acc, s, i) => ({ ...acc, [`:stage${i}`]: s }), {})
  return docClient.queryAllPromise({
    TableName: config.db.recordStagingTable,
    KeyConditionExpression: 'filename = :filename',
    // ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues).join(',')})` }),
    ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues)})` }),
    ExpressionAttributeValues: { ':filename': filename, ...stageValues },
    ConsistentRead: true
  })
}
