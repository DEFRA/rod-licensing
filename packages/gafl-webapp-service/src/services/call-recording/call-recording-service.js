import soapRequest from 'easy-soap-request'
import { XMLBuilder } from 'fast-xml-parser'
import db from 'debug'
const debug = db('webapp:call-recording-service')

export const pauseRecording = async agentEmail => {
  debug('Sending pause recording request to Storm for agentEmail: %s', agentEmail)
  sendRequestToTelesales('pause', agentEmail)
}

export const resumeRecording = async agentEmail => {
  debug('Sending resume recording request to Storm for agentEmail: %s', agentEmail)
  sendRequestToTelesales('resume', agentEmail)
}

const sendRequestToTelesales = async (action, agentEmail) => {
  const url = process.env.TELESALES_ENDPOINT
  const headers = { 'Content-Type': 'text/xml;charset=UTF-8' }
  const xml = await buildXml(action, agentEmail)
  await soapRequest({ url, headers, xml })
}

const buildXml = async (action, agentEmail) => {
  const authUsername = process.env.TELESALES_AUTH_USERNAME
  const authPassword = process.env.TELESALES_AUTH_PASSWORD

  const builder = new XMLBuilder({
    attributesGroupName: '@',
    format: true,
    ignoreAttributes: false,
    suppressBooleanAttributes: false
  })
  const xml = builder.build({
    'soap:Envelope': {
      '@': {
        'xmlns:soap': 'http://www.w3.org/2003/05/soap-envelope',
        'xmlns:cal': 'http://redcentrex.redwoodtech.com/calls'
      },
      'soap:Header': {
        'wsse:Security': {
          '@': {
            'soap:mustUnderstand': 'true',
            'xmlns:wsse': 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd',
            'xmlns:wsu': 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd'
          },
          'wsse:UsernameToken': {
            '@': {
              'wsu:Id': 'UsernameToken-1'
            },
            'wsse:Username': authUsername,
            'wsse:Password': {
              '@': {
                Type: 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText'
              },
              '#text': authPassword
            }
          }
        }
      },
      'soap:Body': {
        'cal:RecordingRequest': {
          'cal:UserID': agentEmail,
          'cal:Record': action,
          'cal:Muted': '0'
        }
      }
    }
  })
  return xml
}
