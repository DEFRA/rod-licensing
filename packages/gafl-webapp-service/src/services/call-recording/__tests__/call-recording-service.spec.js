import soapRequest from 'easy-soap-request'
import db from 'debug'
import { pauseRecording, resumeRecording } from '../call-recording-service'

jest.mock('easy-soap-request')
jest.mock('debug', () => jest.fn(() => jest.fn()))
const { value: debug } = db.mock.results[db.mock.calls.findIndex(c => c[0] === 'webapp:call-recording-service')]

describe('call-recording-service', () => {
  beforeAll(() => {
    process.env.TELESALES_ENDPOINT = 'endpoint'
    process.env.TELESALES_AUTH_USERNAME = 'username'
    process.env.TELESALES_AUTH_PASSWORD = 'password'
  })
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('pauseRecording', () => {
    it('logs that it is requesting to pause the recording', async () => {
      const agentEmail = 'agent@example.com'
      await pauseRecording(agentEmail)
      expect(debug).toHaveBeenCalledWith('Sending pause recording request to Storm for agentEmail: %s', agentEmail)
    })

    it('sends a soap request', async () => {
      await pauseRecording('agent@example.com')
      const expectedArgs = { url: 'endpoint', headers: { 'Content-Type': 'text/xml;charset=UTF-8' } }
      expect(soapRequest).toHaveBeenCalledWith(expect.objectContaining(expectedArgs))
    })

    it('sends a soap request with the correct xml', async () => {
      await pauseRecording('agent@example.com')
      expect(soapRequest.mock.calls[0][0].xml).toMatchSnapshot()
    })
  })

  describe('resumeRecording', () => {
    it('logs that it is requesting to resume the recording', async () => {
      const agentEmail = 'agent@example.com'
      await resumeRecording(agentEmail)
      expect(debug).toHaveBeenCalledWith('Sending resume recording request to Storm for agentEmail: %s', agentEmail)
    })

    it('sends a soap request with the correct headers and endpoint', async () => {
      const expectedArgs = { url: 'endpoint', headers: { 'Content-Type': 'text/xml;charset=UTF-8' } }
      await resumeRecording('agent@example.com')
      expect(soapRequest).toHaveBeenCalledWith(expect.objectContaining(expectedArgs))
    })

    it('sends a soap request with the correct xml', async () => {
      await resumeRecording('agent@example.com')
      expect(soapRequest.mock.calls[0][0].xml).toMatchSnapshot()
    })
  })
})
