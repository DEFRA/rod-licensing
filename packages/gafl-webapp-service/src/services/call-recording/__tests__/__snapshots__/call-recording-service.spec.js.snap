// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call-recording-service pauseRecording sends a soap request with the correct xml 1`] = `
"<soap:Envelope xmlns:soap=\\"http://www.w3.org/2003/05/soap-envelope\\" xmlns:cal=\\"http://redcentrex.redwoodtech.com/calls\\">
  <soap:Header>
    <wsse:Security soap:mustUnderstand=\\"true\\" xmlns:wsse=\\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd\\" xmlns:wsu=\\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd\\">
      <wsse:UsernameToken wsu:Id=\\"UsernameToken-1\\">
        <wsse:Username>username</wsse:Username>
        <wsse:Password Type=\\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText\\">password</wsse:Password>
      </wsse:UsernameToken>
    </wsse:Security>
  </soap:Header>
  <soap:Body>
    <cal:RecordingRequest>
      <cal:UserID>agent@example.com</cal:UserID>
      <cal:Record>pause</cal:Record>
      <cal:Muted>0</cal:Muted>
    </cal:RecordingRequest>
  </soap:Body>
</soap:Envelope>
"
`;

exports[`call-recording-service resumeRecording sends a soap request with the correct xml 1`] = `
"<soap:Envelope xmlns:soap=\\"http://www.w3.org/2003/05/soap-envelope\\" xmlns:cal=\\"http://redcentrex.redwoodtech.com/calls\\">
  <soap:Header>
    <wsse:Security soap:mustUnderstand=\\"true\\" xmlns:wsse=\\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd\\" xmlns:wsu=\\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd\\">
      <wsse:UsernameToken wsu:Id=\\"UsernameToken-1\\">
        <wsse:Username>username</wsse:Username>
        <wsse:Password Type=\\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText\\">password</wsse:Password>
      </wsse:UsernameToken>
    </wsse:Security>
  </soap:Header>
  <soap:Body>
    <cal:RecordingRequest>
      <cal:UserID>agent@example.com</cal:UserID>
      <cal:Record>resume</cal:Record>
      <cal:Muted>0</cal:Muted>
    </cal:RecordingRequest>
  </soap:Body>
</soap:Envelope>
"
`;
